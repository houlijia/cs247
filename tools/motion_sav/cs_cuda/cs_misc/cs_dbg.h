#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
