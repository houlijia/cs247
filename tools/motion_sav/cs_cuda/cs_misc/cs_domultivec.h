#ifndef __CS_DOMULTIVEC_H__
#define __CS_DOMULTIVEC_H__

void
h_do_multi_trnsp_vec ( float *d_input, float *d_output, int *d_R_perm_tbl, int *d_L_perm_tbl,
	int tbl_size ) ;

#ifdef CUDA_OBS 
void
h_do_multi_vec ( float *d_input, float *d_output, int *d_R_perm_tbl, int *d_L_perm_tbl,
	int tbl_size ) ;
#endif 

void
h_do_multi_vec ( float *d_input, float *d_output, float *d_tmp, int *d_R_perm_tbl,
	int *d_L_perm_tbl, int tbl_size, int keep ) ;

#endif 
