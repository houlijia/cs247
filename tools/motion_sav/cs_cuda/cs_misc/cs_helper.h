#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
