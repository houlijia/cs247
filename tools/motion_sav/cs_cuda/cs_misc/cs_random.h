#ifndef __CS_RANDOM_H__
#define __CS_RANDOM_H__

int h_set_random_table ( RndC_uint32 seed, RndC_uint32 *h_in, RndC_uint32 *d_out, int size,
	int slot_one_reserved, int do_init )	 ;

#endif 
