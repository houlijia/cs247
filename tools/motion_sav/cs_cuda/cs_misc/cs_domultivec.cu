#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>
#include "cs_dbg.h"
#include "cs_cuda.h"
#include "cs_helper.h"

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_whm_encode_b.h"
#include "cs_perm_generic.h"
#include "cs_domultivec.h"

#define CUDA_DBG
#define CUDA_DBG1

/*
h_do_multi_vec:

	the transformation is based on walsh-hadamard transform

	d_input : input data
	d_output : output data
	d_tmp : tmp buffer
	R/L_perm_tbl : permutation table
	keep_size : -1 ... keep all 

	NOTE: all d_xxx sizes have to be the same for all table,
	and has to be exactly the size of some power of 2

	tbl_size is the vhtc size, need not to be mod2

	return: data in d_output, d_input is not changed
*/

void
h_do_multi_vec ( float *d_input, float *d_output, float *d_tmp, int *d_R_perm_tbl,
	int *d_L_perm_tbl, int tbl_size, int keep )
{
	int size ;

	size = max_log2( tbl_size ) ;

#ifdef CUDA_DBG 
	printf("%s: din %p dout %p dtmp %p Rperm %p Lperm %p tblsize %d size %d\n",
		__func__, d_input, d_output, d_tmp, d_R_perm_tbl, d_L_perm_tbl, tbl_size, size ) ;
#endif 

	clear_device_mem( d_input + tbl_size, size - tbl_size ) ;
			
#ifdef CUDA_DBG 
	dbg_p_d_data_i("Right perm", d_R_perm_tbl, size ) ; 
	dbg_p_d_data_f("input", d_input, size ) ; 
#endif 

	h_do_permutation_generic_f1( d_input, d_tmp, d_R_perm_tbl, size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f("input after R perm", d_tmp, size ) ; 
#endif 

	cs_whm_measurement_b( d_tmp, size, size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f("input after whm", d_tmp, size ) ; 
	dbg_p_d_data_i("Left perm", d_R_perm_tbl, size ) ; 
#endif 

	h_do_permutation_generic_f2( d_tmp, d_output, d_L_perm_tbl, size ) ;

	if ( keep >= 0 )
		clear_device_mem( d_output + keep, size - keep ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f("input after L perm", d_tmp, size ) ; 
	dbg_p_d_data_f("output after L perm", d_output, size ) ; 
#endif 
}


/*
h_do_multi_trnsp_vec:

	the transformation is based on walsh-hadamard transform

	d_input : input data // 4752 ... filled 0 for the rest
	d_output : output data
	R/L_perm_tbl : permutation table

	size: has to be the same for all table, and has to be exactly the size of some power of 2

return: data in d_input
*/

void
h_do_multi_trnsp_vec ( float *d_input, float *d_output, int *d_R_perm_tbl, int *d_L_perm_tbl,
	int tbl_size )
{
	if ( tbl_size == max_log2( tbl_size ))
	{
		printf("%s: din %p dout %p Rperm %p Lperm %p tblsize %d\n", __func__,
			d_input, d_output, d_R_perm_tbl, d_L_perm_tbl, tbl_size ) ;
	}

#ifdef CUDA_DBG 
	printf("%s: din %p dout %p Rperm %p Lperm %p tblsize %d\n", __func__,
		d_input, d_output, d_R_perm_tbl, d_L_perm_tbl, tbl_size ) ;
#endif 

#ifdef CUDA_DBG 
	dbg_p_d_data_i("Right perm", d_R_perm_tbl, tbl_size ) ; 
	dbg_p_d_data_f("input", d_input, tbl_size ) ; 
#endif 

	h_do_permutation_generic_f1( d_input, d_output, d_L_perm_tbl, tbl_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f("input after R perm", d_output, tbl_size ) ; 
#endif 

	cs_iwhm_measurement_b( d_output, tbl_size, tbl_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f("input after R perm", d_output, tbl_size ) ; 

	dbg_p_d_data_i("Left perm", d_R_perm_tbl, tbl_size ) ; 
#endif 

	h_do_permutation_generic_f2( d_output, d_input, d_R_perm_tbl, tbl_size ) ;

#ifdef CUDA_DBG 
	dbg_p_d_data_f("input after L perm", d_input, tbl_size ) ; 
#endif 
}
