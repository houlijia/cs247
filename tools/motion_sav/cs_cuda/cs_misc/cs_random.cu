#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_cuda.h"
#include "cs_helper.h"

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_random.h"

// #define CUDA_DBG
#define CUDA_DBG1

static RndCState cs_random_state ;

// LDL stop here ... in SensingMatrixWH.makepuermutations
// PL = 1 .. 1 + random(order - 1) 
// PR = random(order) 


// val can be either positive or negative

__global__ void 
d_set_random_table( RndC_uint32 *fp, int size, RndC_uint32 val )
{
	int tid = blockIdx.x*blockDim.x + threadIdx.x;

	while ( tid < size )
	{
		fp[ tid ] += val ;
		tid += CUDA_MAX_THREADS ;
	} 
}

/*
    seed : the seed for random number generation
	h_in: points to the host buffer before download to device, if null, a temp buffer will be
		allocated and deleted when done.
	slot_one_reserved: if set [0] will have 0
*/

int
h_set_random_table ( RndC_uint32 seed, RndC_uint32 *h_in, RndC_uint32 *d_out, int size,
	int slot_one_reserved, int do_init )	
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ;
	RndC_uint32 *h_in_memp = NULL ;
	int i, need_adjustment = 0 ;

	if ( h_in == NULL )
	{
		h_in_memp = ( RndC_uint32 *)malloc ( sizeof ( RndC_uint32 ) * size ) ;
		if ( h_in_memp == NULL )
			return ( 0 ) ;
	} else
		h_in_memp = h_in ;


	if ( do_init )
		init_RndC( &cs_random_state, seed ) ;

	if ( slot_one_reserved )
	{
		h_in_memp[0] = ( RndC_uint32 )0 ;
		randperm1_RndC( &cs_random_state, size - 1, h_in_memp+1 ) ; // from 1 to ( size - 1 )
	} else
	{
		randperm1_RndC( &cs_random_state, size, h_in_memp ) ; // from 1 to ( size )
			// need to subtract one 
		need_adjustment++ ;
	}

	if (( i = cudaMemcpy( d_out, h_in_memp, sizeof ( RndC_uint32 ) * size,
		cudaMemcpyHostToDevice)) != cudaSuccess )
	{
		printf("%s: download cp fail: %d\n", __func__, i ) ;
		
		if ( h_in == NULL )
			free ( h_in_memp ) ;
		return ( 0 ) ;
	}

	if ( need_adjustment )
	{
		h_block_adj ( size, nThreadsPerBlock, &nBlocks ) ;

		d_set_random_table <<< nBlocks, nThreadsPerBlock >>> ( d_out, size, ( RndC_uint32) -1 ) ;

		cudaThreadSynchronize() ;
	}

	if ( h_in == NULL )
		free ( h_in_memp ) ;

	return ( 1 ) ;
}

