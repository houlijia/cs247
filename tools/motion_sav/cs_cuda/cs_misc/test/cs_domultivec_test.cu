#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_random.h"
#include "cs_domultivec.h"

#define CUDA_DBG

#define TST_SIZE 8

float *dp3, *dp1, *dp2, hd[ TST_SIZE ] ;
RndC_uint32 *k1, *k2 ;

int
main( int ac, char *av[] )
{
	int k ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;

	if (( k = cudaMalloc( &dp1, sizeof ( float ) * TST_SIZE * 3 )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 9 ) ;
	}

	if (( k = cudaMalloc( &k1, sizeof ( RndC_uint32 ) * TST_SIZE * 2 )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 9 ) ;
	}

	k2 = k1 + TST_SIZE ;
	dp2 = dp1 + TST_SIZE ;
	dp3 = dp2 + TST_SIZE ;

	for ( k = 0 ; k < TST_SIZE ; k++ )
		hd[k] = 0.5 * k + 1.0 ;

	if (( k = cudaMemcpy( dp1, hd, sizeof ( float ) * TST_SIZE,
		cudaMemcpyHostToDevice)) != cudaSuccess )
	{
		printf("%s: download cp fail: %d\n", __func__, k ) ;
		exit ( 1 ) ;
	}

	dbg_p_d_data_f("before do", dp1, TST_SIZE ) ;  

	if (!h_set_random_table( 0, NULL, k1, TST_SIZE, 0, 1 ))
	{
		printf("%s: k1 fail: \n", __func__ ) ;
		exit ( 1 ) ;
	}

	dbg_p_d_data_i("k1", ( int * )k1, TST_SIZE ) ;  

	if (!h_set_random_table( 1, NULL, k2, TST_SIZE, 0, 1 ))
	{
		printf("%s: k2 fail: \n", __func__ ) ;
		exit ( 1 ) ;
	}

	dbg_p_d_data_i("k2", ( int * )k2, TST_SIZE ) ;  

	h_do_multi_vec ( dp1, dp2, dp3, (int *)k1, (int *)k2, TST_SIZE - 3, -1 ) ;

	dbg_p_d_data_f("after do", dp2, TST_SIZE ) ;	// data in output

	h_do_multi_vec ( dp1, dp2, dp3, (int *)k1, (int *)k2, TST_SIZE - 3, 3 ) ;

	dbg_p_d_data_f("after do", dp2, TST_SIZE ) ;	// data in output
	// do multi transpose vec

	printf("------------------------------------------------------------------------\n") ;

	for ( k = 0 ; k < TST_SIZE ; k++ )
		hd[k] = 0.5 * k + 1.0 ;

	if (( k = cudaMemcpy( dp1, hd, sizeof ( float ) * TST_SIZE,
		cudaMemcpyHostToDevice)) != cudaSuccess )
	{
		printf("%s: download cp fail: %d\n", __func__, k ) ;
		exit ( 1 ) ;
	}

	dbg_p_d_data_f("TRANS:before do", dp1, TST_SIZE ) ;  

	dbg_p_d_data_i("TRANS k1", ( int * )k1, TST_SIZE ) ;  

	dbg_p_d_data_i("TRANS k2", ( int * )k2, TST_SIZE ) ;  

	h_do_multi_trnsp_vec( dp1, dp2, (int *)k1, (int *)k2, TST_SIZE ) ;

	dbg_p_d_data_f("TRANS after do", dp1, TST_SIZE ) ;	// data in input  

}
