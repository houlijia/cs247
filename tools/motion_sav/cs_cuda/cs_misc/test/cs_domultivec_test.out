Fri Jan  2 13:31:51 EST 2015
cs_domultivec_test
---------------------------------------------------
dbg_p_d_data_f: before do size 32 dp 0x1300200000
dbg_pdata_f: before do
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 3.500000
6 -- 4.000000
7 -- 4.500000
dbg_p_d_data_i: k1 size 32 dp 0x1300200200
dbg_pdata_i: k1
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
dbg_p_d_data_i: k2 size 32 dp 0x1300200220
dbg_pdata_i: k2
0 -- 00000002 2
1 -- 00000005 5
2 -- 00000004 4
3 -- 00000006 6
4 -- 00000003 3
5 -- 00000007 7
6 -- 00000000 0
7 -- 00000001 1
h_do_multi_vec: din 0x1300200000 dout 0x1300200020 dtmp 0x1300200040 Rperm 0x1300200200 Lperm 0x1300200220 tblsize 5 size 8
dbg_p_d_data_i: Right perm size 32 dp 0x1300200200
dbg_pdata_i: Right perm
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
dbg_p_d_data_f: input size 32 dp 0x1300200000
dbg_pdata_f: input
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 0.000000
6 -- 0.000000
7 -- 0.000000
h_do_permutation_generic_f1: din 0x1300200000 dout 0x1300200040 perm 0x1300200200 tblsize 8
dbg_p_d_data_f: h_do_permutation_generic_f1 perm before size 32 dp 0x1300200000
dbg_pdata_f: h_do_permutation_generic_f1 perm before
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 0.000000
6 -- 0.000000
7 -- 0.000000
dbg_p_d_data_f: h_do_permutation_generic_f1 perm after size 32 dp 0x1300200040
dbg_pdata_f: h_do_permutation_generic_f1 perm after
0 -- 0.000000
1 -- 0.000000
2 -- 1.500000
3 -- 0.000000
4 -- 3.000000
5 -- 1.000000
6 -- 2.000000
7 -- 2.500000
dbg_p_d_data_f: input after R perm size 32 dp 0x1300200040
dbg_pdata_f: input after R perm
0 -- 0.000000
1 -- 0.000000
2 -- 1.500000
3 -- 0.000000
4 -- 3.000000
5 -- 1.000000
6 -- 2.000000
7 -- 2.500000
dbg_p_d_data_f: input after whm size 32 dp 0x1300200040
dbg_pdata_f: input after whm
0 -- 10.000000
1 -- 3.000000
2 -- -2.000000
3 -- 1.000000
4 -- -7.000000
5 -- 0.000000
6 -- -1.000000
7 -- -4.000000
dbg_p_d_data_i: Left perm size 32 dp 0x1300200200
dbg_pdata_i: Left perm
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
h_do_permutation_generic_f2: din 0x1300200040 dout 0x1300200020 perm 0x1300200220 size 8 tblsize -831385552
dbg_p_d_data_f: input after L perm size 32 dp 0x1300200040
dbg_pdata_f: input after L perm
0 -- 10.000000
1 -- 3.000000
2 -- -2.000000
3 -- 1.000000
4 -- -7.000000
5 -- 0.000000
6 -- -1.000000
7 -- -4.000000
dbg_p_d_data_f: output after L perm size 32 dp 0x1300200020
dbg_pdata_f: output after L perm
0 -- -2.000000
1 -- 0.000000
2 -- -7.000000
3 -- -1.000000
4 -- 1.000000
5 -- -4.000000
6 -- 10.000000
7 -- 3.000000
dbg_p_d_data_f: after do size 32 dp 0x1300200020
dbg_pdata_f: after do
0 -- -2.000000
1 -- 0.000000
2 -- -7.000000
3 -- -1.000000
4 -- 1.000000
5 -- -4.000000
6 -- 10.000000
7 -- 3.000000
h_do_multi_vec: din 0x1300200000 dout 0x1300200020 dtmp 0x1300200040 Rperm 0x1300200200 Lperm 0x1300200220 tblsize 5 size 8
dbg_p_d_data_i: Right perm size 32 dp 0x1300200200
dbg_pdata_i: Right perm
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
dbg_p_d_data_f: input size 32 dp 0x1300200000
dbg_pdata_f: input
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 0.000000
6 -- 0.000000
7 -- 0.000000
h_do_permutation_generic_f1: din 0x1300200000 dout 0x1300200040 perm 0x1300200200 tblsize 8
dbg_p_d_data_f: h_do_permutation_generic_f1 perm before size 32 dp 0x1300200000
dbg_pdata_f: h_do_permutation_generic_f1 perm before
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 0.000000
6 -- 0.000000
7 -- 0.000000
dbg_p_d_data_f: h_do_permutation_generic_f1 perm after size 32 dp 0x1300200040
dbg_pdata_f: h_do_permutation_generic_f1 perm after
0 -- 0.000000
1 -- 0.000000
2 -- 1.500000
3 -- 0.000000
4 -- 3.000000
5 -- 1.000000
6 -- 2.000000
7 -- 2.500000
dbg_p_d_data_f: input after R perm size 32 dp 0x1300200040
dbg_pdata_f: input after R perm
0 -- 0.000000
1 -- 0.000000
2 -- 1.500000
3 -- 0.000000
4 -- 3.000000
5 -- 1.000000
6 -- 2.000000
7 -- 2.500000
dbg_p_d_data_f: input after whm size 32 dp 0x1300200040
dbg_pdata_f: input after whm
0 -- 10.000000
1 -- 3.000000
2 -- -2.000000
3 -- 1.000000
4 -- -7.000000
5 -- 0.000000
6 -- -1.000000
7 -- -4.000000
dbg_p_d_data_i: Left perm size 32 dp 0x1300200200
dbg_pdata_i: Left perm
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
h_do_permutation_generic_f2: din 0x1300200040 dout 0x1300200020 perm 0x1300200220 size 8 tblsize -831385552
dbg_p_d_data_f: input after L perm size 32 dp 0x1300200040
dbg_pdata_f: input after L perm
0 -- 10.000000
1 -- 3.000000
2 -- -2.000000
3 -- 1.000000
4 -- -7.000000
5 -- 0.000000
6 -- -1.000000
7 -- -4.000000
dbg_p_d_data_f: output after L perm size 32 dp 0x1300200020
dbg_pdata_f: output after L perm
0 -- -2.000000
1 -- 0.000000
2 -- -7.000000
3 -- 0.000000
4 -- 0.000000
5 -- 0.000000
6 -- 0.000000
7 -- 0.000000
dbg_p_d_data_f: after do size 32 dp 0x1300200020
dbg_pdata_f: after do
0 -- -2.000000
1 -- 0.000000
2 -- -7.000000
3 -- 0.000000
4 -- 0.000000
5 -- 0.000000
6 -- 0.000000
7 -- 0.000000
------------------------------------------------------------------------
dbg_p_d_data_f: TRANS:before do size 32 dp 0x1300200000
dbg_pdata_f: TRANS:before do
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 3.500000
6 -- 4.000000
7 -- 4.500000
dbg_p_d_data_i: TRANS k1 size 32 dp 0x1300200200
dbg_pdata_i: TRANS k1
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
dbg_p_d_data_i: TRANS k2 size 32 dp 0x1300200220
dbg_pdata_i: TRANS k2
0 -- 00000002 2
1 -- 00000005 5
2 -- 00000004 4
3 -- 00000006 6
4 -- 00000003 3
5 -- 00000007 7
6 -- 00000000 0
7 -- 00000001 1
h_do_multi_trnsp_vec: din 0x1300200000 dout 0x1300200020 Rperm 0x1300200200 Lperm 0x1300200220 tblsize 8
h_do_multi_trnsp_vec: din 0x1300200000 dout 0x1300200020 Rperm 0x1300200200 Lperm 0x1300200220 tblsize 8
dbg_p_d_data_i: Right perm size 32 dp 0x1300200200
dbg_pdata_i: Right perm
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
dbg_p_d_data_f: input size 32 dp 0x1300200000
dbg_pdata_f: input
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 3.500000
6 -- 4.000000
7 -- 4.500000
h_do_permutation_generic_f1: din 0x1300200000 dout 0x1300200020 perm 0x1300200220 tblsize 8
dbg_p_d_data_f: h_do_permutation_generic_f1 perm before size 32 dp 0x1300200000
dbg_pdata_f: h_do_permutation_generic_f1 perm before
0 -- 1.000000
1 -- 1.500000
2 -- 2.000000
3 -- 2.500000
4 -- 3.000000
5 -- 3.500000
6 -- 4.000000
7 -- 4.500000
dbg_p_d_data_f: h_do_permutation_generic_f1 perm after size 32 dp 0x1300200020
dbg_pdata_f: h_do_permutation_generic_f1 perm after
0 -- 4.000000
1 -- 4.500000
2 -- 1.000000
3 -- 3.000000
4 -- 2.000000
5 -- 1.500000
6 -- 2.500000
7 -- 3.500000
dbg_p_d_data_f: input after R perm size 32 dp 0x1300200020
dbg_pdata_f: input after R perm
0 -- 4.000000
1 -- 4.500000
2 -- 1.000000
3 -- 3.000000
4 -- 2.000000
5 -- 1.500000
6 -- 2.500000
7 -- 3.500000
dbg_p_d_data_f: input after R perm size 32 dp 0x1300200020
dbg_pdata_f: input after R perm
0 -- 2.750000
1 -- -0.375000
2 -- 0.250000
3 -- 0.375000
4 -- 0.375000
5 -- -0.250000
6 -- 0.875000
7 -- 0.000000
dbg_p_d_data_i: Left perm size 32 dp 0x1300200200
dbg_pdata_i: Left perm
0 -- 00000005 5
1 -- 00000002 2
2 -- 00000006 6
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000000 0
6 -- 00000001 1
7 -- 00000003 3
h_do_permutation_generic_f2: din 0x1300200020 dout 0x1300200000 perm 0x1300200200 size 8 tblsize -831385552
dbg_p_d_data_f: input after L perm size 32 dp 0x1300200000
dbg_pdata_f: input after L perm
0 -- -0.250000
1 -- 0.250000
2 -- 0.875000
3 -- 0.000000
4 -- 0.375000
5 -- 2.750000
6 -- -0.375000
7 -- 0.375000
dbg_p_d_data_f: TRANS after do size 32 dp 0x1300200000
dbg_pdata_f: TRANS after do
0 -- -0.250000
1 -- 0.250000
2 -- 0.875000
3 -- 0.000000
4 -- 0.375000
5 -- 2.750000
6 -- -0.375000
7 -- 0.375000
